dbt/include/spark/macros/materializations/table.sql

```diff
@@ -41,7 +41,17 @@
 # --- Autogenerated dbt materialization code. --- #
 dbt = dbtObj(spark.table)
 df = model(dbt, spark)
-df.write.mode("overwrite").format("delta").saveAsTable("{{ target_relation }}")
+
+# make sure pandas exists
+import importlib.util
+package_name = 'pandas'
+if importlib.util.find_spec(package_name):
+    import pandas
```

Comment:
Is this line required?

From what I see, in databricks we might usually want to load `pyspark.pandas` rather than `pandas` (and I don't think it is required here)

Reply:
@b-per I believe it _is_ required in order to do test the type: `isinstance(df, pandas.core.frame.DataFrame)`. I tried commenting out the single line with the import, and it did not work.

However, digging into your question did uncover something else...

### Goal
For the `py_write_table()` macro, we want the return type to be `pyspark.sql.dataframe.DataFrame`.

### Potential input types

There are three different data types we expect <sup>[1](#internal_references)</sup> it to be:
1. `pyspark.sql.dataframe.DataFrame` (PySpark DataFrame)
1. `pandas.core.frame.DataFrame` (Pandas DataFrame)
1. `pyspark.pandas.frame.DataFrame` (Pandas-on-Spark DataFrame)

Are we handling each of the three cases?
1. ✅ For the first case, it is already a `pyspark.sql.dataframe.DataFrame`, so no conversion necessary.
1. ✅ For the second case, this PR infers if it is a `pandas.core.frame.DataFrame` and converts it using `spark.createDataFrame()` if so.
1. ❌ However, the third case isn't handled yet! We can handle it by calling `to_spark()` on it <sup>[2](#internal_references)</sup>.

I'll update this PR to include the third case.

---

<a name="internal_references"></a>

1. We could proactively raise an exception if it falls outside of these expected types (rather than sending it off to the database where it will fall on its face and emit an unhelpful error message).
1. Thank you to @Adricarpin for this handy resource: https://towardsdatascience.com/run-pandas-as-fast-as-spark-f5eefe780c45#64b7

Reply:
Yep, my bad, I thought `isinstance` was comparing with a string and not an object type.

Reply:
@b-per Very glad you asked, because now we have a more robust implementation as a result!